### PR TITLE
Spike/dsapi to sapint

### PR DESCRIPTION
**Pull request by bogdanadrianmarc**
Since this gem is currently only used by the LR apps, the changes are based around that. Originally, one of the apps would generate a query via the `QueryGenerator` class and that would be sent off to DSAPI. This PR changes that, so after the query is generated, before being sent off, the query is converted to SAPINT style and then sent off to a SAPINT instance. After that, the JSON in the response is converted to DSAPI JSON response style and returned to the LR app. Currently only the `explain` functionality is turned off, this will change after SAPINT implements that functionality

### CODE REVIEW

**Review by ijdickinson — CHANGES_REQUESTED** [review]
It would be much more idiomatic Ruby to define methods with `def` in the normal way, rather than have an array of lambdas. That would also allow you to factor out some of the common code (like dealing with `@id` hashes), and make it easier to call one method from another. To dispatch the conversions, Ruby's `case` statement would be a good solution.


In general, it would be better to create or transform arrays and then join them, rather than do string bashing. So instead of:

```ruby
      '@oneof' => lambda { |attribute, values|
        result = ''
        values.each do |value|
          result += "#{VALUE_FILTERS['@eq'].call(attribute, value)}&"
        end
        result.slice(0...-1)
      },
```

it would be better to do:

```ruby
      '@oneof' => lambda { |attribute, values|
        values.map do |value|
          VALUE_FILTERS['@eq'].call(attribute, value)
        end .join('&')
      },
```

The code to reformat the response from SapiNT to DSAPI format should be separated out, so that we can easily remove it if the view formatting can in future be done in SapiNT as I suggested.

We should really have tests for the new code

File: lib/data_services_api/dataset.rb [hunk 1]
```diff
@@ -49,15 +49,54 @@ def structure
...
+
+    # Converts SAPINT returned JSON format to DSAPI returned JSON format
+    def to_dsapi_response(sapint_response)
+      sapint_response['items'].map do |value|
+        to_dsapi_item(value)
+      end
+    end
+
+    def to_dsapi_item(sapint_item)
+      dsapi_item = {}
```
Comment (ijdickinson) [inline lib/data_services_api/dataset.rb:79]: [map](https://ruby-doc.org/core-3.0.1/Enumerable.html#method-i-map) is defined on `Enumerable`, so can be used directly on  a `Hash`:

```ruby
def to_dsapi_item(sapint_item)
  sapint_item.map { |k, v| [k, to_dsapi_json(k, v)] }.to_h
end
```

File: lib/data_services_api/dataset.rb [hunk 2]
```diff
@@ -49,15 +49,54 @@ def structure
...
+      sapint_item.each do |key, value|
+        dsapi_json = to_dsapi_json(key, value)
+        dsapi_item.merge!(dsapi_json)
+      end
+
+      dsapi_item
+    end
+
+    def to_dsapi_json(sapint_key, sapint_value)
+      dataset_name = data_api.split('/').slice(-1)
```
Comment (ijdickinson) [inline lib/data_services_api/dataset.rb:90]: I think it would be clearer to use a regex with a capture group here

File: lib/data_services_api/dataset.rb [hunk 3]
```diff
@@ -49,15 +49,54 @@ def structure
...
+
+      return { sapint_key => sapint_value } if sapint_key == '@id'
+      return { "#{dataset_name}:#{sapint_key}" => sapint_value } unless sapint_value.is_a?(Hash)
+
+      dsapi_json = {}
```
Comment (ijdickinson) [inline lib/data_services_api/dataset.rb:95]: See note above about using `map` directly on a hash

File: lib/data_services_api/dataset.rb [hunk 4]
```diff
@@ -49,15 +49,54 @@ def structure
...
+      converter = SapiNTConverter.new(query.to_json)
+      sapi_query_params = converter.to_sapint_query
+      explain_url = "#{data_api}/explain"
+      { sparql: service.api_get_text(explain_url, sapi_query_params) }
+    end
+
+    private
+
+    # Converts SAPINT returned JSON format to DSAPI returned JSON format
+    def to_dsapi_response(sapint_response)
```
Comment (ijdickinson) [inline lib/data_services_api/dataset.rb:72]: Converting to the DsAPI syntactic form is not a responsibility of a `Dataset` per se. I think it should be a separate class. That will also make it easier to write the tests

File: lib/data_services_api/sapint_converter.rb [hunk 5]
```diff
@@ -0,0 +1,147 @@
+# frozen_string_literal: true
+
+module DataServicesApi
+  class SapiNTConverter
+    def initialize(dsapi_query)
+      @dsapi_query = JSON.parse(dsapi_query)
+    end
+
+    # Converts a DSAPI query to SAPINT query
+    def to_sapint_query
```
Comment (ijdickinson) [inline lib/data_services_api/sapint_converter.rb:10]: I thought after our discussion on Thursday you were going to leave the URL parameters as a `Hash`, so that Faraday would handle them correctly? Or is `to_sapint_query` not used any more, in which case it should be deleted

File: lib/data_services_api/sapint_converter.rb [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+      result = @dsapi_query.map do |key, value|
+        sapint_query(key, value)
+      end.compact.join('&')
+      sanitize_result(result)
+    end
+
+    private
+
+    def sapint_query(key, value)
```
Comment (ijdickinson) [inline lib/data_services_api/sapint_converter.rb:19]: This is much clearer than the previous use of lambdas. :+1:

File: lib/data_services_api/sapint_converter.rb [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+    def sort(values)
+      values.map do |value|
+        sort_prop = value.key?('@up') ? "+#{value['@up']}" : "-#{value['@down']}"
+        "_sort=#{sort_prop}"
+      end
+    end
+
+    def count(value)
+      '_count=@id' if value == true
```
Comment (ijdickinson) [inline lib/data_services_api/sapint_converter.rb:42]: You only need `if value` here

File: lib/data_services_api/sapint_converter.rb [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+    def limit(value)
+      "_limit=#{value}"
+    end
+
+    def offset(value)
+      "_offset=#{value}"
+    end
+
+    def andd(values)
```
Comment (ijdickinson) [inline lib/data_services_api/sapint_converter.rb:53]: The logic here is a bit hard to follow with the use of `value`, `values` and `[]` indexing. Can you please refactor to make the logic a bit clearer?

File: lib/data_services_api/sapint_converter.rb [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+      "#{attribute}=#{value['@value']}"
+    end
+
+    def oneof(attribute, values)
+      values.map do |value|
+        eq(attribute, value)
+      end
+    end
+
+    def ge(attribute, value)
```
Comment (ijdickinson) [inline lib/data_services_api/sapint_converter.rb:94]: There seems to be an opportunity to DRY -up `ge`, `gt`, etc as they are all very similar

File: lib/data_services_api/sapint_converter.rb [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    def search(attribute, value)
+      property = remove_prefix(value['@property'])
+      search_text = sanitize_search(value['@value'])
+      "searchPath=#{attribute}&search-#{property}=#{search_text}"
+    end
+
+    def sanitize_search(search_text)
+      search_text
+        .gsub(' AND ', ' ')
+        .gsub('( ', '')
```
Comment (ijdickinson) [inline lib/data_services_api/sapint_converter.rb:127]: You can do both the parens in one step:

```
irb(main):005:0> q = "( foo AND bar )"
=> "( foo AND bar )"
irb(main):006:0> q.gsub(/(\( | \))/, '')
=> "foo AND bar"
irb(main):007:0> 
```

File: lib/data_services_api/sapint_converter.rb [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+        .gsub(' )', '')
+    end
+
+    def remove_prefix(value)
+      value.split(':')[1] unless value.slice(0...4) == 'http'
```
Comment (ijdickinson) [inline lib/data_services_api/sapint_converter.rb:132]: I think it would be clearer to say `unless value.match?(/^http/)`

File: lib/data_services_api/sapint_converter.rb [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+    end
+
+    # Removes duplicate searchPaths and constructs JSON from params array
+    def sanitize_result(result)
```
Comment (ijdickinson) [inline lib/data_services_api/sapint_converter.rb:136]: If the URL params were left as a hash, you wouldn't need to split the string. As we discussed on Thursday, it's not a good idea to generate a string too early in the process

**Review by ijdickinson — COMMENTED** [review]
I think the translation from the DsAPI structure to the SapiNT form is much clearer now. :+1: 

There are a few more comments to address, and I would like to see some reasonable amount of test coverage please.

File: lib/data_services_api/service.rb [hunk 13]
```diff
@@ -3,7 +3,7 @@
 module DataServicesApi
   # Denotes the encapsulated DataServicesAPI service
   class Service
-    DEFAULT_URL = 'http://localhost:8080/dsapi'
+    DEFAULT_URL = 'http://localhost:8080'
```
[the remark this answers is not in the document]
Comment (bogdanadrianmarc) [reply lib/data_services_api/service.rb:6]: The user can currently pass that in the `initialize` method for the `Service` class, do you mean to remove the default fallback value and always expect the API URL from the user? That would be set in each LR app container anyways, so I think it's ok to have a fallback value for development purposes at least

File: spec/data_services_api/aspect_spec.rb [hunk 14]
```diff
@@ -1,53 +1,55 @@
...
-  it 'should have a range type' do
-    _(@aspect1.range_type).must_match(%r{http://www.w3.org/2001/XMLSchema#})
-  end
-
-  it 'should have a reference to the service object' do
-    _(@aspect1.service).wont_be_nil
-    _(@aspect1.service.url).wont_be_nil
-  end
-end
+# This feature is currently unavailable from SAPINT
```
Comment (ijdickinson) [inline spec/data_services_api/aspect_spec.rb:3]: I think it will be OK to just delete this spec. It will remain in the git history anyway

File: spec/data_services_api/dataset_spec.rb [hunk 15]
```diff
@@ -12,29 +12,33 @@
...
-    structure = @dataset.structure
-    _(structure).wont_be_nil
-    _(structure).must_respond_to :size
-
-    structure.each do |aspect|
-      _(aspect.id).wont_be_nil
-    end
-    # end
-  end
+  # This feature is currently unavailable from SAPINT
```
Comment (ijdickinson) [inline spec/data_services_api/dataset_spec.rb:15]: I would just delete these tests, and leave a comment to explain that to future readers

File: lib/data_services_api/dsapi_response_converter.rb [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+    end
+
+    def to_dsapi_json(sapint_key, sapint_value)
+      dataset_name = @data_api[/[a-z]*$/]
+
+      return { sapint_key => sapint_value } if sapint_key == '@id'
+      return { "#{dataset_name}:#{sapint_key}" => sapint_value } unless sapint_value.is_a?(Hash)
+
+      sapint_value.map do |key, value|
+        ["#{dataset_name}:#{sapint_key}#{key == '@id' ? '' : key.capitalize}", value]
```
Comment (ijdickinson) [inline lib/data_services_api/dsapi_response_converter.rb:32]: When is the `key.capitalize` call used? I couldn't see any examples of this in the tests you've written

**Review by ijdickinson — APPROVED** [review]
I think we are pretty close now, thanks. I've made a few minor suggestions, but nothing that blocks merging.

I **am** a bit concerned by the `santiize_search` call. Looking at the tests, it's clear that the "un-sanitized" is in Lucene syntax. I can't remember now why we needed to pass raw Lucene queries instead of search fragments, but we did have problems with Lucene not always giving us the results we wanted. We may have to switch to not-sanitising the search, and passing the strings as "raw" Lucene, which I believe that we can do with SapiNT. However, what I propose for now is that we merge these changes as-is, but then run some extensive tests on the apps when we have a new test environment up and running. The `hmlr-feature-tests` repo will also be helpful in this respect.

File: lib/data_services_api/dsapi_response_converter.rb (hunk 16 above)
Comment (bogdanadrianmarc) [reply lib/data_services_api/dsapi_response_converter.rb:32]: For example if we have:
```
...
propertyAddress: {
    ...
    county: ...
    ...
}
...
```
this should become:
```
...
propertyAddressCounty: ...
...
```
I'll add a test to try to simplify what's going on for a future reader 👍
